nemo/lightning/run/plugins.py

```diff
@@ -0,0 +1,121 @@
+from pathlib import Path
+from typing import Callable, Optional
+
+import nemo_run as run
+import yaml
+from nemo_run.core.serialization.yaml import YamlSerializer
+from pytorch_lightning import Callback
+from pytorch_lightning.loggers import WandbLogger
+
+from nemo.lightning.pytorch.callbacks import NsysCallback, PreemptionCallback
+
+
+def _merge_callbacks(partial: run.Partial, callbacks: list[run.Config[Callback]]):
+    if hasattr(partial, "trainer"):
+        if hasattr(partial.trainer, "callbacks"):
+            for callback in callbacks:
+                if callback not in partial.trainer.callbacks:
+                    partial.trainer.callbacks.append(callback)
+        else:
+            partial.trainer.callbacks = copy.deepcopy(callbacks)
+
+
+@dataclass(kw_only=True)
+class PreemptionPlugin(run.Plugin):
+    callbacks: list[run.Config[Callback]] = field(default_factory=lambda: [run.Config(PreemptionCallback)])
+
+    def setup(self, task: run.Partial | run.Script, executor: run.Executor):
+        if isinstance(task, run.Script):
+            logging.warning(
+                f"The {self.__class__.__name__} will have no effect on the task as it's an instance of run.Script"
+            )
+            return
+
+        if isinstance(executor, run.SlurmExecutor):
+            # Sends a SIGTERM 5 minutes before hitting time limit
+            executor.signal = "TERM@300"
+
+        _merge_callbacks(task, callbacks=self.callbacks)
+
+
+def nsys_plugin(
+    partial: run.Partial,
+    executor: run.Executor,
+    start_step: int,
+    end_step: int,
+    ranks: Optional[list] = None,
+    nsys_trace: Optional[list[str]] = None,
+):
+    nsys_callback = run.Config(
+        NsysCallback,
+        start_step=start_step,
+        end_step=end_step,
+        ranks=ranks or [0],
+    )
+    callbacks: list[run.Config[Callback]] = [nsys_callback]  # type: ignore
+    _merge_callbacks(partial, callbacks=callbacks)
+
+    launcher = executor.get_launcher()
+    launcher.nsys_profile = True
+    launcher.nsys_trace = nsys_trace or ["nvtx", "cuda"]
+
+
+@dataclass(kw_only=True)
+class NsysPlugin(run.Plugin):
+    start_step: int
+    end_step: int
+    ranks: Optional[list[int]] = None
+    nsys_trace: Optional[list[str]] = None
+
+    def setup(self, task: run.Partial | run.Script, executor: run.Executor):
+        if isinstance(task, run.Partial):
+            nsys_callback = run.Config(
+                NsysCallback,
+                start_step=self.start_step,
+                end_step=self.end_step,
+                ranks=self.ranks or [0],
+            )
+            callbacks: list[run.Config[Callback]] = [nsys_callback]  # type: ignore
+            _merge_callbacks(task, callbacks=callbacks)
+
+        launcher = executor.get_launcher()
+        launcher.nsys_profile = True
+        launcher.nsys_trace = self.nsys_trace or ["nvtx", "cuda"]
+
+
+@dataclass(kw_only=True)
+class WandbPlugin(run.Plugin):
+    name: str
```

Comment:
Add documentation asking users to set WANDB_API_KEY and some info on how to get the key

Reply:
Added docstrings for the plugins in https://github.com/NVIDIA/NeMo/pull/10223/commits/51dc9d07e73fdc689f20ae6ee42d788096ae9608